dataset/src/main/kotlin/org/jetbrains/kotlinx/dl/dataset/preprocessor/Preprocessing.kt

```diff
@@ -51,45 +49,25 @@ public class Preprocessing {
     }
 
     internal fun handleFile(file: File): Pair<FloatArray, ImageShape> {
-        val filename = file.name
         var image = imagePreprocessingStage.load.fileToImage(file)
-        var shape = imagePreprocessingStage.load.imageShape!!
-
-        // if both nulls write a warning to logs about possible mismatch
-        if (imagePreprocessingStage.isCropInitialized) {
-            val (croppedImage, croppedShape) = imagePreprocessingStage.crop.apply(image, shape)
-            image = croppedImage
-            shape = croppedShape
-        }
-
-        if (imagePreprocessingStage.isRotateInitialized) {
-            val (rotatedImage, rotatedShape) = imagePreprocessingStage.rotate.apply(image, shape)
-            image = rotatedImage
-            shape = rotatedShape
-        }
 
-        if (imagePreprocessingStage.isResizeInitialized) {
-            val (resizedImage, resizedShape) = imagePreprocessingStage.resize.apply(image, shape)
-            image = resizedImage
-            shape = resizedShape
-        }
-
-        if (imagePreprocessingStage.isSaveInitialized) {
-            imagePreprocessingStage.save.imageToFile(filename, image, shape)
+        for (operation in imagePreprocessingStage.operations) {
+            if (operation is Save) {
```

Comment:
Looks like we have two sort of transformation: intermediate and start/final.
What do you think about Save op? Is it an imageTransformation or we should have a separate block for it at the top level?
load {
}
transformImage {
}
save {
}

Reply:
But on other side, what if we want to save checkpoints after each imageTransformation (or display it)

Reply:
> But on other side, what if we want to save checkpoints after each imageTransformation (or display it)

Yes, that what I was thinking.

One of the possible ways it could look like:

```
resize {
    outputWidth = 400
    outputHeight = 400
    interpolation = InterpolationType.NEAREST
    save {
        dirLocation = preprocessedImagesDirectory
    }
}
```